File: ruby/lib/fizzy/security.rb
# frozen_string_literal: true

require "uri"

module Fizzy
  # Security helpers for URL validation, message truncation, and header redaction.
  # Used across the SDK to enforce HTTPS, prevent SSRF, and protect sensitive data.
  module Security
    MAX_ERROR_MESSAGE_BYTES = 500
    MAX_RESPONSE_BODY_BYTES = 50 * 1024 * 1024 # 50 MB
    MAX_ERROR_BODY_BYTES = 1 * 1024 * 1024      # 1 MB

    def self.truncate(str, max = MAX_ERROR_MESSAGE_BYTES)
      return str if str.nil? || str.bytesize <= max

      max <= 3 ? str.byteslice(0, max) : str.byteslice(0, max - 3) + "..."
    end

    def self.require_https!(url, label = "URL")
      uri = URI.parse(url.to_s)
      raise UsageError.new("#{label} must use HTTPS: #{url}") unless uri.scheme&.downcase == "https"
    rescue URI::InvalidURIError
      raise UsageError.new("Invalid #{label}: #{url}")
    end

    def self.same_origin?(a, b)
      ua = URI.parse(a)
      ub = URI.parse(b)
      return false if ua.scheme.nil? || ub.scheme.nil?

      ua.scheme.downcase == ub.scheme.downcase &&
        normalize_host(ua) == normalize_host(ub)
    rescue URI::InvalidURIError
      false
    end

    def self.resolve_url(base, target)
      URI.join(base, target).to_s
    rescue URI::InvalidURIError
      target
    end

    def self.normalize_host(uri)
      host = uri.host&.downcase
      port = uri.port
      return host if port.nil?
      return host if uri.scheme&.downcase == "https" && port == 443
      return host if uri.scheme&.downcase == "http" && port == 80

      "#{host}:#{port}"
    end

    def self.check_body_size!(body, max, label = "Response")
      return if body.nil?

      if body.bytesize > max
        raise Fizzy::APIError.new(
          "#{label} body too large (#{body.bytesize} bytes, max #{max})"
        )
      end
    end

    def self.localhost?(url)
      uri = URI.parse(url.to_s)
      host = uri.host&.downcase
      return false if host.nil?

      host == "localhost" ||
        host == "127.0.0.1" ||
        host == "::1" ||
        host == "[::1]" ||
        host.end_with?(".localhost")
    rescue URI::InvalidURIError
      false
    end

    def self.require_https_unless_localhost!(url, label = "URL")
      return if localhost?(url)

      require_https!(url, label)
    end

    # Headers that contain sensitive values and should be redacted.
    SENSITIVE_HEADERS = %w[
      authorization
      cookie
      set-cookie
      x-csrf-token
    ].freeze

    # Returns a copy of the headers with sensitive values replaced by "[REDACTED]".
    #
    # @param headers [Hash] the headers hash (case-insensitive keys)
    # @return [Hash] a new hash with sensitive values redacted
    def self.redact_headers(headers)
      result = {}
      headers.each do |key, value|
        result[key] = SENSITIVE_HEADERS.include?(key.to_s.downcase) ? "[REDACTED]" : value
      end
      result
    end
  end
end

File: typescript/src/security.ts
/**
 * Security utilities for the Fizzy SDK.
 *
 * Provides helpers for safely logging HTTP requests without exposing
 * sensitive information like tokens and cookies.
 */

const SENSITIVE_HEADERS = [
  "authorization",
  "cookie",
  "set-cookie",
  "x-csrf-token",
];

/**
 * Returns a copy of the headers with sensitive values replaced by "[REDACTED]".
 */
export function redactHeaders(headers: Headers): Record<string, string> {
  const result: Record<string, string> = {};

  headers.forEach((value, key) => {
    const lowerKey = key.toLowerCase();
    if (SENSITIVE_HEADERS.includes(lowerKey)) {
      result[key] = "[REDACTED]";
    } else {
      result[key] = value;
    }
  });

  return result;
}

/**
 * Returns a copy of the header record with sensitive values replaced by "[REDACTED]".
 */
export function redactHeadersRecord(
  headers: Record<string, string>
): Record<string, string> {
  const result: Record<string, string> = {};

  for (const [key, value] of Object.entries(headers)) {
    const lowerKey = key.toLowerCase();
    if (SENSITIVE_HEADERS.includes(lowerKey)) {
      result[key] = "[REDACTED]";
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Checks if a hostname represents localhost for development/testing purposes.
 */
export function isLocalhost(hostname: string): boolean {
  const normalized = hostname.toLowerCase();
  return (
    normalized === "localhost" ||
    normalized === "127.0.0.1" ||
    normalized === "::1" ||
    normalized.endsWith(".localhost")
  );
}
